ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,132 @@
+// Its value is the number of spans to sample per second.
+// Spans that matched the rules but exceeded the rate limit are not sampled.
+type traceRulesSampler struct {
+	rules      []SamplingRule // the rules to match spans with
+	globalRate float64        // a rate to apply when no rules match a span
+	limiter    *rateLimiter   // used to limit the volume of spans sampled
+}
+
+// newTraceRulesSampler configures a *traceRulesSampler instance using the given set of rules.
+// Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
+func newTraceRulesSampler(rules []SamplingRule) *traceRulesSampler {
+	return &traceRulesSampler{
+		rules:      rules,
+		globalRate: globalSampleRate(),
+		limiter:    newRateLimiter(),
+	}
+}
+
+func (rs *traceRulesSampler) enabled() bool {
+	return len(rs.rules) > 0 || !math.IsNaN(rs.globalRate)
+}
+
+// apply uses the sampling rules to determine the sampling rate for the
+// provided span. If the rules don't match, and a default rate hasn't been
+// set using DD_TRACE_SAMPLE_RATE, then it returns false and the span is not
+// modified.
+func (rs *traceRulesSampler) apply(span *span) bool {
+	if !rs.enabled() {
+		// short path when disabled
+		return false
+	}
+
+	var matched bool
+	rate := rs.globalRate
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			matched = true
+			rate = rule.Rate
+			break
+		}
+	}
+	if !matched && math.IsNaN(rate) {
+		// no matching rule or global rate, so we want to fall back
+		// to priority sampling
+		return false
+	}
+
+	rs.applyRate(span, rate, time.Now())
+	return true
+}
+
+func (rs *traceRulesSampler) applyRate(span *span, rate float64, now time.Time) {
+	span.SetTag(keyRulesSamplerAppliedRate, rate)
+	if !sampledByRate(span.TraceID, rate) {
+		span.setSamplingPriority(ext.PriorityUserReject, samplernames.RuleRate, rate)
+		return
+	}
+
+	sampled, rate := rs.limiter.allowOne(now)
+	if sampled {
+		span.setSamplingPriority(ext.PriorityUserKeep, samplernames.RuleRate, rate)
+	} else {
+		span.setSamplingPriority(ext.PriorityUserReject, samplernames.RuleRate, rate)
+	}
+	span.SetTag(keyRulesSamplerLimiterRate, rate)
+}
+
+// limit returns the rate limit set in the rules sampler, controlled by DD_TRACE_RATE_LIMIT, and
+// true if rules sampling is enabled. If not present it returns math.NaN() and false.
+func (rs *traceRulesSampler) limit() (float64, bool) {
+	if rs.enabled() {
+		return float64(rs.limiter.limiter.Limit()), true
+	}
+	return math.NaN(), false
+}
+
+// defaultRateLimit specifies the default trace rate limit used when DD_TRACE_RATE_LIMIT is not set.
+const defaultRateLimit = 100.0
+
+// newRateLimiter returns a rate limiter which restricts the number of traces sampled per second.
+// This defaults to 100.0. The DD_TRACE_RATE_LIMIT environment variable may override the default.
+func newRateLimiter() *rateLimiter {
+	limit := defaultRateLimit
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v != "" {
```

Comment:
```suggestion
	if v := os.Getenv("DD_TRACE_RATE_LIMIT"); v != ""  {
```

Reply:
nit: also slight preference to name it something like `envVal` to be clearer, but that's optional

Reply:
I'd keep `v` for a few reasons:
* It's what we've generally used.
* There is no need for brevity here as this variable is only used twice two lines below where it is defined.
* In this case `v` is quite commonly used in Go.

Generally I prefer to follow the guidelines from https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#naming-things. I think the general guideline for naming should be: "_the farther away a variable is used, the longer and more descriptive its name should be_".

In code, we want to make the algorithm evident, and not the purpose of variables. That should be made clear at the place of definition. Repetitively reading a variable that is only used a few times but contains one or more words will obstruct the visibility of the algorithm and is unnecessary.